### PR TITLE
feat: add basic support openApi 3.1

### PR DESCRIPTION
**Pull request by AlexVarchuk**
fixes https://github.com/Redocly/openapi-cli/issues/276

### CODE REVIEW

**Review by andriyl — APPROVED** [review]

File: packages/core/src/types/oas3_1.ts [hunk 1]
```diff
@@ -0,0 +1,586 @@
...
+    openapi: null,
+    info: 'Info',
+    servers: listOf('Server'),
+    security: listOf('SecurityRequirement'),
+    tags: listOf('Tag'),
+    externalDocs: 'ExternalDocs',
+    paths: 'PathMap',
+    webhooks: 'Webhooks',
+    components: 'Components',
+    'x-webhooks': 'WebhooksMap',
```
Comment (RomanHotsiy) [inline packages/core/src/types/oas3_1.ts:17]: No need for x-webhooks

Comment (RomanHotsiy) [inline packages/core/src/types/oas3_1.ts:17]: No need for x-webhooks in 3.1
```suggestion
```

File: packages/core/src/types/oas3_1.ts [hunk 2]
```diff
@@ -0,0 +1,586 @@
...
+const DefinitionRoot: NodeType = {
+  properties: {
+    openapi: null,
+    info: 'Info',
+    servers: listOf('Server'),
+    security: listOf('SecurityRequirement'),
+    tags: listOf('Tag'),
+    externalDocs: 'ExternalDocs',
+    paths: 'PathMap',
+    webhooks: 'Webhooks',
```
Comment (RomanHotsiy) [inline packages/core/src/types/oas3_1.ts:15]:
```suggestion
    webhooks: 'WebhooksMap',
```

File: packages/core/src/types/oas3_1.ts [hunk 3]
```diff
@@ -0,0 +1,586 @@
...
+    components: 'Components',
+    'x-webhooks': 'WebhooksMap',
+  },
+  required: ['openapi', 'info'],
+};
+
+const Tag: NodeType = {
```
Comment (RomanHotsiy) [inline packages/core/src/types/oas3_1.ts:22]: can we reuse shared types from oas3.0 ?

File: packages/core/src/types/oas3_1.ts [hunk 4]
```diff
@@ -0,0 +1,586 @@
...
+const PathMap: NodeType = {
+  properties: {},
+  additionalProperties: (_value: any, key: string) =>
+    key.startsWith('/') ? 'PathItem' : undefined,
+};
+
+const Webhooks: NodeType = {
+  properties: {},
+  additionalProperties: {},
+};
```
Comment (RomanHotsiy) [inline packages/core/src/types/oas3_1.ts:130]:
```suggestion
```

File: packages/core/src/types/oas3_1.ts [hunk 5]
```diff
@@ -0,0 +1,586 @@
...
+    uniqueItems: { type: 'boolean' },
+    maxProperties: { type: 'number' },
+    minProperties: { type: 'number' },
+    required: { type: 'array', items: { type: 'string' } }, // like array
+    enum: { type: 'array' },
+    type: (value: any) => {
+      if (Array.isArray(value)) {
+        return { type: 'array', items: { enum: ['object', 'array', 'string', 'number', 'integer', 'boolean', 'null'] } }
+      } else {
+      return {
```
Comment (RomanHotsiy) [inline packages/core/src/types/oas3_1.ts:376]: formatting

File: packages/core/src/types/oas3_1.ts [hunk 6]
```diff
@@ -0,0 +1,586 @@
...
+    operationRef: { type: 'string' },
+    operationId: { type: 'string' },
+    parameters: null, // TODO: figure out how to describe/validate this
+    requestBody: null, // TODO: figure out how to describe/validate this
+    description: { type: 'string' },
+    server: 'Server',
+  },
+};
+
+const Schema: NodeType = {
```
Comment (RomanHotsiy) [inline packages/core/src/types/oas3_1.ts:352]: you should add more fields here, new json schema draft supports `if`, `else`, `const`, etc.

File: packages/core/src/types/oas3_1.ts [hunk 7]
```diff
@@ -0,0 +1,586 @@
...
+    implicit: 'ImplicitFlow',
+    password: 'PasswordFlow',
+    clientCredentials: 'ClientCredentials',
+    authorizationCode: 'AuthorizationCode',
+  },
+};
+
+const SecurityScheme: NodeType = {
+  properties: {
+    type: { enum: ['apiKey', 'http', 'oauth2', 'openIdConnect'] },
```
Comment (RomanHotsiy) [inline packages/core/src/types/oas3_1.ts:510]: There is one more added I think, `mutualTls`

**Review by RomanHotsiy — CHANGES_REQUESTED** [review]
Left a bunch of comments.

**Review by Marshevskyy — COMMENTED** [review]
good job Alex!

it would be amazing to add some tests which can cover this functionality,
we do not need to cover some implementation, 
but we should cover behavior

ideally we should somehow pass as an input OAS 3.1 and got it valid in linter....

**Comment by AlexVarchuk** [comment]
> good job Alex!
> 
> it would be amazing to add some tests which can cover this functionality,
> we do not need to cover some implementation,
> but we should cover behavior
> 
> ideally we should somehow pass as an input OAS 3.1 and got it valid in linter....

Good idea. I will try to implement it.

**Comment by AlexVarchuk** [comment]
@RomanHotsiy @Marshevskyy Please check it again. Thanks.

**Review by Marshevskyy — APPROVED** [review]
Alex! well done! 
tests 💪🏼 

but before merge, I would like to get approval from @RomanHotsiy

**Comment by timbunce** [comment]
Just a note from a random bystander to say Thanks Alex! Looking forward to using this.

**Review by RomanHotsiy — APPROVED** [review]
Good job 👏

**Pull request merged**